### PR TITLE
Refactor component feature and version labels

### PR DESCRIPTION
**Pull request by chrismuellner**
# Description
The components (activegate and oneagent) managed by the operator should have their version included in the labels.

Following labels/annotations were changed:
- Set `app.kubernetes.io/name` to `oneagent` or `activegate`
- Set `app.kubernetes.io/version` to oneagent or activegate version
- Replaced `component.dynatrace.com/feature` with `app.kubernetes.io/component` for the oneagent mode or activegate capability
- Removed version annotation on activegate/oneagent pods

## How can this be tested?
Deploy a dynakube with any oneagent mode and any activegate capability enabled.

Expected labels after this change:
- oneagent:
```yaml
  labels:
    app.kubernetes.io/name: oneagent
    app.kubernetes.io/component: cloudnativefullstack
    app.kubernetes.io/version: 1.241.0.20220413-171953
    app.kubernetes.io/created-by: dynakube
    app.kubernetes.io/managed-by: dynatrace-operator
    
```

```yaml
  selector:
    matchLabels:
      app.kubernetes.io/name: oneagent
      app.kubernetes.io/created-by: dynakube
      app.kubernetes.io/managedy-by: dynatrace-operator
      
```
- activegate:
```yaml
  labels:
    app.kubernetes.io/name: activegate
    app.kubernetes.io/component: activegate
    app.kubernetes.io/version: 1.241.0.20220414-004007
    app.kubernetes.io/created-by: dynakube
    app.kubernetes.io/managed-by: dynatrace-operator
    
```

```yaml
  selector:
    matchLabels:
      app.kubernetes.io/name: activegate
      app.kubernetes.io/created-by: dynakube
      app.kubernetes.io/managedy-by: dynatrace-operator
      
```
- other resources (e.g. secret) that are connected to oneagent/activegate but shouldn't change between versions
```yaml
labels:
    app.kubernetes.io/component: activegate
    app.kubernetes.io/created-by: dynakube
    app.kubernetes.io/name: dynatrace-operator
    
```

## Checklist
- [x] Unit tests have been updated/added
- [x] PR is labeled accordingly

### CODE REVIEW

File: src/controllers/dynakube/oneagent/daemonset/daemonset.go [hunk 1]
```diff
@@ -128,13 +127,15 @@ func appendHostIdArgument(result *appsv1.DaemonSet, source string) {
 func (dsInfo *builderInfo) BuildDaemonSet() (*appsv1.DaemonSet, error) {
 	instance := dsInfo.instance
 	podSpec := dsInfo.podSpec()
+
+	podLabels := kubeobjects.NewComponentLabels(instance.Name, kubeobjects.OneAgentComponentLabel,
+		dsInfo.deploymentType, instance.Status.OneAgent.Version)
 	labels := kubeobjects.MergeLabels(
```
Comment (mjgrzybek) [inline src/controllers/dynakube/oneagent/daemonset/daemonset.go:133]:
```suggestion
	daemonSetLabels := kubeobjects.MergeLabels(
```

File: src/controllers/dynakube/oneagent/oneagent_reconciler.go [hunk 2]
```diff
@@ -149,11 +149,14 @@ func (r *OneAgentReconciler) getDesiredDaemonSet(dkState *status.DynakubeState)
 	return dsDesired, nil
 }
 
+// getPods loads pods managed by the operator
```
Comment (0sewa0) [inline src/controllers/dynakube/oneagent/oneagent_reconciler.go:152]:
```suggestion
// getPods loads oneagent pods managed by the operator
```

or just rename the function to `getOneAgentPods` and remove the comment as it clear enough at that point (I prefer this option 😛 )

File: src/kubeobjects/labels.go [hunk 3]
```diff
@@ -21,6 +22,69 @@ const (
...
+	}
+}
+
+func NewPodLabels(createdBy string, component ComponentLabelValue) *PodLabels {
+	return NewComponentLabels(createdBy, component, "", "")
+}
+
+// BuildLabels produces a set of labels that
+// include versions of operator and component
+// and component feature, if set
```
Comment (0sewa0) [inline src/kubeobjects/labels.go:61]:
```suggestion
// BuildLabels produces a set of labels that
// contain the matchLabels and 
// extra labels that can change overtime
// or depending the features that are enabled.
```

Nitpick: I think the original comment is a bit too specific, if I want to see the exact labels created then I will look at the code.

The important thing that needs to be mentioned IMO is how its different from `BuildMatchLabels`, which is that it contains labels that can change.
By not being super specific, we will not have to change the comment when we add a new label.

File: src/kubeobjects/labels.go [hunk 4]
```diff
@@ -2,24 +2,111 @@ package kubeobjects
...
+// include oneagent/activegate mode and version
+func (labels *AppLabels) BuildLabels() map[string]string {
+	labelsMap := labels.BuildMatchLabels()
+	labelsMap[AppVersionLabel] = labels.Version
+	labelsMap[AppComponentLabel] = labels.Component
+	return labelsMap
+}
+
+// BuildMatchLabels produces a set of labels that
+// don't change when switching operator versions
```
Comment (meik99) [inline src/kubeobjects/labels.go:92]:
```suggestion
// doesn't change when switching operator versions
```

File: src/kubeobjects/labels.go [hunk 5]
```diff
@@ -2,24 +2,111 @@ package kubeobjects
...
+func (labels *coreMatchLabels) BuildMatchLabels() map[string]string {
+	return map[string]string{
+		AppNameLabel:      labels.Name,
+		AppCreatedByLabel: labels.CreatedBy,
+		AppComponentLabel: labels.Component,
+	}
+}
+
+// BuildMatchLabels produces a set of labels that
+// don't change when switching oneagent or activegate mode
```
Comment (meik99) [inline src/kubeobjects/labels.go:102]:
```suggestion
// doesn't change when switching oneagent or activegate mode
```

**Review by meik99 — CHANGES_REQUESTED** [review]

File: src/agproxysecret/secret.go [hunk 6]
```diff
@@ -48,14 +48,15 @@ func (agProxySecretGenerator *ActiveGateProxySecretGenerator) GenerateForDynakub
...
+	coreLabels := kubeobjects.NewCoreLabels(dynakube.Name, kubeobjects.ActiveGateComponentLabel)
 	return kubeobjects.CreateOrUpdateSecretIfNotExists(
 		agProxySecretGenerator.client,
 		agProxySecretGenerator.apiReader,
 		BuildProxySecretName(),
 		agProxySecretGenerator.namespace,
 		data,
-		labels,
+		coreLabels.BuildMatchLabels(),
 		corev1.SecretTypeOpaque,
```
[the remark this answers is not in the document]
Comment (chrismuellner) [reply src/agproxysecret/secret.go:60]: out of scope of this PR

File: src/kubeobjects/labels.go [hunk 7]
```diff
@@ -2,24 +2,111 @@ package kubeobjects
...
+// BuildLabels produces a set of labels that
+// include operator version
+func (labels *CoreLabels) BuildLabels() map[string]string {
+	labelsMap := labels.BuildMatchLabels()
+	labelsMap[AppVersionLabel] = labels.Version
+	return labelsMap
+}
+
+// BuildLabels produces a set of labels that
+// include oneagent/activegate mode and version
```
Comment (0sewa0) [inline src/kubeobjects/labels.go:83]:
```suggestion
// include values that can change overtime
```

File: src/kubeobjects/labels.go (hunk 4 above)
Comment (0sewa0) [inline src/kubeobjects/labels.go:92]:
```suggestion
// BuildMatchLabels creates labels that
// don't include values that change overtime
```

File: src/kubeobjects/labels.go (hunk 5 above)
Comment (0sewa0) [inline src/kubeobjects/labels.go:102]:
```suggestion
// BuildMatchLabels creates labels that
// don't include values that change overtime
```

**Review by 0sewa0 — APPROVED** [review]

**Pull request merged**